File: python/src/dflow/solana/transactions.py
"""Solana transaction utilities for DFlow SDK."""

import asyncio
import base64
import time
from typing import Literal

from solana.rpc.api import Client
from solders.keypair import Keypair
from solders.transaction import VersionedTransaction

from dflow.types import TransactionConfirmation

Commitment = Literal["processed", "confirmed", "finalized"]


def sign_and_send_transaction(
    connection: Client,
    transaction_base64: str,
    signer: Keypair,
) -> str:
    """Sign and send a base64-encoded transaction to the Solana network.

    Deserializes the transaction, signs it with the provided keypair,
    and sends it to the network.

    Args:
        connection: Solana RPC connection
        transaction_base64: Base64-encoded transaction (from DFlow API responses)
        signer: Keypair to sign the transaction with

    Returns:
        Transaction signature

    Example:
        >>> from solana.rpc.api import Client
        >>> from solders.keypair import Keypair
        >>> from dflow import sign_and_send_transaction
        >>>
        >>> connection = Client("https://api.mainnet-beta.solana.com")
        >>> keypair = Keypair.from_bytes(secret_key)
        >>>
        >>> order = dflow.orders.get_order(...)
        >>> signature = sign_and_send_transaction(connection, order.transaction, keypair)
        >>> print(f"Transaction sent: {signature}")
    """
    # Decode base64 transaction
    transaction_buffer = base64.b64decode(transaction_base64)
    transaction = VersionedTransaction.from_bytes(transaction_buffer)

    # Sign with keypair
    transaction.sign([signer])

    # Send transaction
    result = connection.send_transaction(transaction)

    return str(result.value)


def wait_for_confirmation(
    connection: Client,
    signature: str,
    commitment: Commitment = "confirmed",
    timeout_ms: int = 60000,
) -> TransactionConfirmation:
    """Wait for a transaction to be confirmed on-chain.

    Polls the network until the transaction reaches the desired confirmation
    level or times out.

    Args:
        connection: Solana RPC connection
        signature: Transaction signature to wait for
        commitment: Desired confirmation level (default: 'confirmed')
        timeout_ms: Maximum time to wait in milliseconds (default: 60000)

    Returns:
        Confirmation details including slot and status

    Raises:
        Exception: If transaction fails or times out

    Example:
        >>> from dflow import wait_for_confirmation
        >>>
        >>> confirmation = wait_for_confirmation(connection, signature, "confirmed")
        >>> print(f"Confirmed at slot {confirmation.slot}")
    """
    start_time = time.time()

    while (time.time() - start_time) * 1000 < timeout_ms:
        response = connection.get_signature_statuses([signature])
        status = response.value[0] if response.value else None

        if status:
            if status.err:
                raise Exception(f"Transaction failed: {status.err}")

            confirmation_status = status.confirmation_status
            if confirmation_status:
                status_str = str(confirmation_status).lower()
                if (
                    status_str == "finalized"
                    or (commitment == "confirmed" and status_str in ("confirmed", "finalized"))
                    or (commitment == "processed" and status_str in ("processed", "confirmed", "finalized"))
                ):
                    return TransactionConfirmation(
                        signature=signature,
                        slot=status.slot,
                        confirmation_status=status_str,  # type: ignore
                        err=str(status.err) if status.err else None,
                    )

        time.sleep(2)

    raise TimeoutError(f"Transaction confirmation timeout after {timeout_ms}ms")


def sign_send_and_confirm(
    connection: Client,
    transaction_base64: str,
    signer: Keypair,
    commitment: Commitment = "confirmed",
) -> TransactionConfirmation:
    """Sign, send, and wait for confirmation in one call.

    Convenience function that combines sign_and_send_transaction and
    wait_for_confirmation into a single operation.

    Args:
        connection: Solana RPC connection
        transaction_base64: Base64-encoded transaction (from DFlow API responses)
        signer: Keypair to sign the transaction with
        commitment: Desired confirmation level (default: 'confirmed')

    Returns:
        Confirmation details including signature, slot, and status

    Raises:
        Exception: If transaction fails or times out

    Example:
        >>> from solana.rpc.api import Client
        >>> from solders.keypair import Keypair
        >>> from dflow import DFlowClient, sign_send_and_confirm, USDC_MINT
        >>>
        >>> dflow = DFlowClient()
        >>> connection = Client("https://api.mainnet-beta.solana.com")
        >>> keypair = Keypair.from_bytes(secret_key)
        >>>
        >>> # Get a swap transaction
        >>> swap = dflow.swap.create_swap(
        ...     input_mint=USDC_MINT,
        ...     output_mint=yes_mint,
        ...     amount=1000000,
        ...     slippage_bps=50,
        ...     user_public_key=str(keypair.pubkey()),
        ... )
        >>>
        >>> # Sign, send, and wait for confirmation
        >>> result = sign_send_and_confirm(connection, swap.swap_transaction, keypair)
        >>> print(f"Transaction confirmed: {result.signature}")
        >>> print(f"Slot: {result.slot}")
    """
    signature = sign_and_send_transaction(connection, transaction_base64, signer)
    return wait_for_confirmation(connection, signature, commitment)


async def wait_for_confirmation_async(
    connection: Client,
    signature: str,
    commitment: Commitment = "confirmed",
    timeout_ms: int = 60000,
) -> TransactionConfirmation:
    """Async version of wait_for_confirmation.

    Args:
        connection: Solana RPC connection
        signature: Transaction signature to wait for
        commitment: Desired confirmation level (default: 'confirmed')
        timeout_ms: Maximum time to wait in milliseconds (default: 60000)

    Returns:
        Confirmation details including slot and status

    Raises:
        Exception: If transaction fails or times out
    """
    start_time = time.time()

    while (time.time() - start_time) * 1000 < timeout_ms:
        response = connection.get_signature_statuses([signature])
        status = response.value[0] if response.value else None

        if status:
            if status.err:
                raise Exception(f"Transaction failed: {status.err}")

            confirmation_status = status.confirmation_status
            if confirmation_status:
                status_str = str(confirmation_status).lower()
                if (
                    status_str == "finalized"
                    or (commitment == "confirmed" and status_str in ("confirmed", "finalized"))
                    or (commitment == "processed" and status_str in ("processed", "confirmed", "finalized"))
                ):
                    return TransactionConfirmation(
                        signature=signature,
                        slot=status.slot,
                        confirmation_status=status_str,  # type: ignore
                        err=str(status.err) if status.err else None,
                    )

        await asyncio.sleep(2)

    raise TimeoutError(f"Transaction confirmation timeout after {timeout_ms}ms")

File: packages/typescript/src/solana/transactions.ts
import {
  Connection,
  VersionedTransaction,
  Keypair,
  type Commitment,
  type SignatureStatus,
} from '@solana/web3.js';
import type { TransactionConfirmation } from '../types/index.js';

/**
 * Sign and send a base64-encoded transaction to the Solana network.
 *
 * Deserializes the transaction, signs it with the provided keypair,
 * and sends it to the network.
 *
 * @param connection - Solana RPC connection
 * @param transactionBase64 - Base64-encoded transaction (from DFlow API responses)
 * @param signer - Keypair to sign the transaction with
 * @returns Transaction signature
 *
 * @example
 * ```typescript
 * import { Connection, Keypair } from '@solana/web3.js';
 * import { signAndSendTransaction } from 'dflow-sdk';
 *
 * const connection = new Connection('https://api.mainnet-beta.solana.com');
 * const keypair = Keypair.fromSecretKey(secretKey);
 *
 * const order = await dflow.orders.getOrder({ ... });
 * const signature = await signAndSendTransaction(connection, order.transaction, keypair);
 * console.log(`Transaction sent: ${signature}`);
 * ```
 */
export async function signAndSendTransaction(
  connection: Connection,
  transactionBase64: string,
  signer: Keypair
): Promise<string> {
  const transactionBuffer = Buffer.from(transactionBase64, 'base64');
  const transaction = VersionedTransaction.deserialize(transactionBuffer);

  transaction.sign([signer]);

  const signature = await connection.sendTransaction(transaction, {
    skipPreflight: false,
    preflightCommitment: 'confirmed',
  });

  return signature;
}

/**
 * Wait for a transaction to be confirmed on-chain.
 *
 * Polls the network until the transaction reaches the desired confirmation
 * level or times out.
 *
 * @param connection - Solana RPC connection
 * @param signature - Transaction signature to wait for
 * @param commitment - Desired confirmation level (default: 'confirmed')
 * @param timeoutMs - Maximum time to wait in milliseconds (default: 60000)
 * @returns Confirmation details including slot and status
 * @throws Error if transaction fails or times out
 *
 * @example
 * ```typescript
 * import { waitForConfirmation } from 'dflow-sdk';
 *
 * const confirmation = await waitForConfirmation(connection, signature, 'confirmed');
 * console.log(`Confirmed at slot ${confirmation.slot}`);
 * ```
 */
export async function waitForConfirmation(
  connection: Connection,
  signature: string,
  commitment: Commitment = 'confirmed',
  timeoutMs: number = 60000
): Promise<TransactionConfirmation> {
  const startTime = Date.now();

  while (Date.now() - startTime < timeoutMs) {
    const response = await connection.getSignatureStatus(signature);
    const status: SignatureStatus | null = response.value;

    if (status) {
      if (status.err) {
        throw new Error(`Transaction failed: ${JSON.stringify(status.err)}`);
      }

      const confirmationStatus = status.confirmationStatus;
      if (
        confirmationStatus === 'finalized' ||
        (commitment === 'confirmed' && confirmationStatus === 'confirmed') ||
        (commitment === 'processed' && confirmationStatus === 'processed')
      ) {
        return {
          signature,
          slot: status.slot,
          confirmationStatus: confirmationStatus ?? 'processed',
          err: status.err,
        };
      }
    }

    await sleep(2000);
  }

  throw new Error(`Transaction confirmation timeout after ${timeoutMs}ms`);
}

/**
 * Sign, send, and wait for confirmation in one call.
 *
 * Convenience function that combines {@link signAndSendTransaction} and
 * {@link waitForConfirmation} into a single operation.
 *
 * @param connection - Solana RPC connection
 * @param transactionBase64 - Base64-encoded transaction (from DFlow API responses)
 * @param signer - Keypair to sign the transaction with
 * @param commitment - Desired confirmation level (default: 'confirmed')
 * @returns Confirmation details including signature, slot, and status
 * @throws Error if transaction fails or times out
 *
 * @example
 * ```typescript
 * import { Connection, Keypair } from '@solana/web3.js';
 * import { DFlowClient, signSendAndConfirm, USDC_MINT } from 'dflow-sdk';
 *
 * const dflow = new DFlowClient();
 * const connection = new Connection('https://api.mainnet-beta.solana.com');
 * const keypair = Keypair.fromSecretKey(secretKey);
 *
 * // Get a swap transaction
 * const swap = await dflow.swap.createSwap({
 *   inputMint: USDC_MINT,
 *   outputMint: yesMint,
 *   amount: 1000000,
 *   slippageBps: 50,
 *   userPublicKey: keypair.publicKey.toBase58(),
 * });
 *
 * // Sign, send, and wait for confirmation
 * const result = await signSendAndConfirm(connection, swap.transaction, keypair);
 * console.log(`Transaction confirmed: ${result.signature}`);
 * console.log(`Slot: ${result.slot}`);
 * ```
 */
export async function signSendAndConfirm(
  connection: Connection,
  transactionBase64: string,
  signer: Keypair,
  commitment: Commitment = 'confirmed'
): Promise<TransactionConfirmation> {
  const signature = await signAndSendTransaction(connection, transactionBase64, signer);
  return waitForConfirmation(connection, signature, commitment);
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
